### PR TITLE
Re-work type mapping API

### PR DESCRIPTION
**Pull request by ajcvickers**
This change introduces a new interface for type mapping that effectively replaces IRelationalTypeMapper. This means:
* Type mapping is now first-class at the non-relational level, and all value conversion has been moved here
* RelationalTypeMapper has been reverted to it's 2.0 state so that providers don't have to figure out how to deal with all the changes
* A fallback mechanism is in place that uses IRelationalTypeMapper until a provider implements a new type mapper
  * IRelationalCoreTypeMapper cannot handle everything, but this keeps current code working until providers update
* The concept of a common type mapping implementation for all relational providers has been dropped. This really didn't work out--SQLite, Postgres, MySQL, SQL Server, and Oracle all have different requirements and in effect implement type mapping in different ways, so the base implementation provides little value.
  * Providers now just override one method and can do so in whatever way makes sense.

The new type-mapping API is not fully implemented for SQL Server or SQLite yet since the main focus here is on getting the fallback API to work for 2.1.

### CODE REVIEW

File: src/EFCore.Relational/Metadata/Conventions/Internal/RelationalConventionSetBuilderDependencies.cs [hunk 1]
```diff
@@ -80,22 +90,15 @@ public sealed class RelationalConventionSetBuilderDependencies
...
         /// </summary>
         public RelationalConventionSetBuilderDependencies(
-            [NotNull] IRelationalTypeMapper typeMapper,
+            [NotNull] IRelationalCoreTypeMapper coreTypeMapper,
             [CanBeNull] IDiagnosticsLogger<DbLoggerCategory.Model> logger,
             [CanBeNull] ICurrentDbContext currentContext,
-            [CanBeNull] IDbSetFinder setFinder)
-            : this(setFinder, currentContext, logger, typeMapper)
+            [CanBeNull] IDbSetFinder setFinder,
+            [CanBeNull] IRelationalTypeMapper _ = null) // Only needed for D.I. to resolve this constructor
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Conventions/Internal/RelationalConventionSetBuilderDependencies.cs]: Why? The other constructor has 3 parameters

File: src/EFCore.Relational/Storage/RelationalCoreTypeMapperBase.cs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    ///         This type is typically used by database providers (and other extensions). It is generally
+    ///         not used in application code.
+    ///     </para>
+    /// </summary>
+    public abstract class RelationalCoreTypeMapperBase : CoreTypeMapperBase, IRelationalCoreTypeMapper
+    {
+        /// <summary>
+        ///     TODO
+        /// </summary>
+        protected RelationalCoreTypeMapperBase(
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Storage/RelationalCoreTypeMapperBase.cs]: Doc comment?

File: src/EFCore.Relational/Metadata/Conventions/Internal/RelationalConventionSetBuilderDependencies.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Metadata/Conventions/Internal/RelationalConventionSetBuilderDependencies.cs]:
```
System.InvalidOperationException : Unable to activate type 'Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.RelationalConventionSetBuilderDependencies'. The following constructors are ambiguous:
Void .ctor(Microsoft.EntityFrameworkCore.Storage.IRelationalCoreTypeMapper, Microsoft.EntityFrameworkCore.Diagnostics.IDiagnosticsLogger`1[Microsoft.EntityFrameworkCore.DbLoggerCategory+Model], Microsoft.EntityFrameworkCore.Internal.ICurrentDbContext, Microsoft.EntityFrameworkCore.Internal.IDbSetFinder)
Void .ctor(Microsoft.EntityFrameworkCore.Storage.IRelationalTypeMapper, Microsoft.EntityFrameworkCore.Internal.ICurrentDbContext, Microsoft.EntityFrameworkCore.Internal.IDbSetFinder)
```

File: src/EFCore.Relational/Storage/RelationalCoreTypeMapperBase.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Storage/RelationalCoreTypeMapperBase.cs]: Done

**Comment by ajcvickers** [comment]
Merged

**Pull request closed without merging**